zkevm-circuits/src/evm_circuit/execution/address.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::N_BYTES_ACCOUNT_ADDRESS,
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            from_bytes, CachedRegion, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::CallContextFieldTag,
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{Field, ToLittleEndian};
+use halo2_proofs::plonk::Error;
+use std::convert::TryInto;
+
+#[derive(Clone, Debug)]
+pub(crate) struct AddressGadget<F> {
+    same_context: SameContextGadget<F>,
+    address: RandomLinearCombination<F, N_BYTES_ACCOUNT_ADDRESS>,
+}
+
+impl<F: Field> ExecutionGadget<F> for AddressGadget<F> {
+    const NAME: &'static str = "ADDRESS";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ADDRESS;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let address = cb.query_rlc();
+
+        // Lookup callee address in call context.
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::CalleeAddress,
+            from_bytes::expr(&address.cells),
+        );
+
+        cb.stack_push(address.expr());
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(2.expr()),
+            program_counter: Delta(1.expr()),
+            stack_pointer: Delta((-1).expr()),
+            gas_left: Delta(-OpcodeId::ADDRESS.constant_gas_cost().expr()),
+            ..Default::default()
+        };
+
+        let opcode = cb.query_cell();
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+
+        Self {
+            same_context,
+            address,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let address = block.rws[step.rw_indices[1]].stack_value();
```

Comment:
Might be worth asserting that the `address` is in fact the call's `callee_address`:

```suggestion
        // you will need `use eth_types::ToAddress;` trait
        let address = block.rws[step.rw_indices[1]].stack_value();
        debug_assert_eq!(call.callee_address, address.to_address());
```